oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java

```diff
@@ -633,20 +655,62 @@ void write(ChannelOutboundInvoker out, ChannelPromise promise, ByteBufAllocator
                 headerBuf.writeBoolean(isComposited);
             }
             headerBuf.writeInt(compositedElementCount);
-
-            // write to channel
+            // track content length which will be used for decompress
+            headerBuf.writeBoolean(false);
+            headerBuf.writeInt(originLength);
             out.write(headerBuf);
+            // write content to channel
             out.write(compositedElementLengths);
             for (int i = 0; i < columnCount; i++) {
                 out.write(dataBuffers[i]);
             }
-            if (columnCount > 0) {

```

Comment:
should keep this. `columnCount` could be 0

Reply:
So if columnCount =0, we just need to write the headerBuf? no need to write compositedElementLengths as well as dataBuffers?

Reply:
yes.

Reply:
Refact the code. Please take review again.